backend/experiment/rules/tests/test_speech2song.py

```diff
@@ -70,3 +70,30 @@ def test_next_round_serialization(self):
         self.assertEqual(type(serialized), list)
         for s in serialized:
             self.assertEqual(type(s), dict)
+
+    def test_runthrough(self):
+        speech2song = self.session.block_rules()
+        speech2song.n_trials_per_block = 2
+        for i in range(self.block.rounds - 1):
+            actions = speech2song.next_round(self.session)
+            feedback_trial = next((a for a in actions if a.__dict__.get('feedback_form')))
+            result = Result.objects.get(pk=feedback_trial.feedback_form.form[0].result_id)
+            result.score = 42
+            result.save()
+            if i == 0:
+                self.assertEqual(self._number_of_sound_trials(actions), speech2song.n_presentations)
+            elif i == 1:
+                self.assertIsInstance(actions[0], Explainer)
+                self.assertEqual(self._number_of_sound_trials(actions), 1)
+            elif i == self.block.rounds - 1:
+                self.assertIsInstance(actions[0], Final)
+            elif i == speech2song.n_trials_per_block * 2 * 3 + 1:
+                self.assertIsInstance(actions[0], Explainer)
+            elif i % 2 == 0:
+                self.assertEqual(self._number_of_sound_trials(actions), speech2song.n_presentations)
+            elif i % 2 == 1:
+                self.assertEqual(self._number_of_sound_trials(actions), 1)
```

Comment:
Is this a test of the complete flow? I was thinking of trying to make this happen for more experiments. Simulate the frontend and so on, almost like an E2E test but faster

Reply:
Yes, I thought to go through all the rounds which should be theoretically there. Indeed, I think we can test a lot of the stuff that can go wrong in this way. Perhaps something to discuss: where do we expect problems to happen? Perhaps the e2e tests could also work with pretty dumb mock rules, which only present each action (with potential flavours) once? And testing the logic of the rules files in this way?